statistics/dml_update.go

```diff
@@ -0,0 +1,272 @@
+	BktDeltas   map[int]*BktDelta
+}
+
+// BktDelta stands for the delta info for a bucket.
+type BktDelta struct {
+	Count int64
+	Bound *types.Datum // Bound is the new lower bound of the bucket. For the last bucket, it is also used to store the new upper bound of the bucket.
+}
+
+// RunDynamicUpdate indicates if this TiDB server run dynamic update.
+var RunDynamicUpdate = false
+
+// UpdateTableDelta updates the table delta according to the changed row.
+func (delta *TableDelta) UpdateTableDelta(sc *stmtctx.StatementContext, h *Handle, tableInfo *model.TableInfo, row []types.Datum, deltaCount, count int64) {
+	delta.Delta += deltaCount
+	delta.Count += count
+	if !RunDynamicUpdate {
+		return
+	}
+	table := h.GetTableStats(tableInfo.ID)
+	if table.Pseudo {
+		return
+	}
+	// Update the primary key's histogram.
+	if tableInfo.PKIsHandle {
+		pk := tableInfo.GetPkColInfo()
+		c, ok := table.Columns[pk.ID]
+		if ok && pk.State == model.StatePublic && pk.Offset >= 0 && pk.Offset < len(row) {
+			if delta.PKDelta == nil {
+				delta.PKID = pk.ID
+				delta.PKDelta = &HistDelta{}
+			}
+			delta.PKDelta.updateHistDelta(sc, &c.Histogram, &row[pk.Offset], deltaCount)
+		}
+	}
+	// Update the indices' histogram.
+	for _, idxInfo := range tableInfo.Indices {
+		if idxInfo.State != model.StatePublic {
+			continue
+		}
+		idx, ok := table.Indices[idxInfo.ID]
+		if !ok {
+			continue
+		}
+		value := encodeIndexValue(h.ctx.GetSessionVars().StmtCtx, idxInfo, row)
+		if value == nil {
+			continue
+		}
+		if delta.IdxDeltas == nil {
+			delta.IdxDeltas = make(map[int64]*HistDelta)
+		}
+		if delta.IdxDeltas[idxInfo.ID] == nil {
+			delta.IdxDeltas[idxInfo.ID] = &HistDelta{}
+		}
+		idxDelta := delta.IdxDeltas[idxInfo.ID]
+		idxDelta.updateHistDelta(sc, &idx.Histogram, value, deltaCount)
+	}
+}
+
+func encodeIndexValue(sc *stmtctx.StatementContext, idxInfo *model.IndexInfo, row []types.Datum) *types.Datum {
+	values := make([]types.Datum, 0, len(idxInfo.Columns))
+	for _, col := range idxInfo.Columns {
+		if col.Offset < 0 || col.Offset >= len(row) {
+			return nil
+		}
+		values = append(values, row[col.Offset])
+	}
+	b, err := codec.EncodeKey(sc, nil, values...)
+	if err != nil {
+		terror.Log(err)
+		return nil
+	}
+	d := types.NewBytesDatum(b)
+	return &d
+}
+
+func (delta *BktDelta) updateBktBound(sc *stmtctx.StatementContext, value *types.Datum, isUpper bool) {
+	if delta.Bound == nil {
+		delta.Bound = value
+		return
+	}
+	cmp, err := delta.Bound.CompareDatum(sc, value)
+	if err != nil {
+		terror.Log(err)
+		return
+	}
+	if (isUpper && cmp == -1) || (!isUpper && cmp == 1) {
+		delta.Bound = value
+	}
+}
+
+func (delta *HistDelta) updateHistDelta(sc *stmtctx.StatementContext, hg *Histogram, value *types.Datum, deltaCount int64) {
+	if hg.Len() == 0 {
+		return
+	}
+	// The histogram version has changed, so we need to migrate the old delta info to the new histogram.
+	if hg.LastUpdateVersion != delta.HistVersion {
```

Comment:
I don't quite understand this logic. What's wrong if current histogram is newer than delta?

Reply:
If the histogram is newer than delta, the bucket boundaries may be completely changed, so that the delta count cannot be directly used anymore.

Reply:
So the version is updated only when the boundaries has been changed ?

Reply:
Nope, that is hard to guarantee. We only know that the histogram is changed.